File: backend/app/services/privacy_engine.py
"""
Advanced Privacy Engine for Synthetic Data
Implements differential privacy, GDPR compliance, and intelligent privacy protection
"""

import numpy as np
from app.utils.optional_imports import pd, PANDAS_AVAILABLE
from typing import Dict, List, Any, Optional, Tuple, Union
import logging
from dataclasses import dataclass
from enum import Enum
import hashlib
import json
from datetime import datetime, timedelta
# Import privacy libraries with graceful fallbacks
try:
    import opendp as dp
    from opendp.measurements import make_base_discrete_laplace, make_base_gaussian
    from opendp.transformations import make_identity, make_clamp, make_bounded_resize
    OPENDP_AVAILABLE = True
except ImportError:
    # Fallback implementations when OpenDP is not available
    OPENDP_AVAILABLE = False
    dp = None

try:
    from diffprivlib import BudgetAccountant, mechanisms
    DIFFPRIVLIB_AVAILABLE = True
except ImportError:
    # Create mock classes when diffprivlib is not available
    DIFFPRIVLIB_AVAILABLE = False
    class BudgetAccountant:
        def __init__(self): pass
    class mechanisms: pass

import warnings

from app.core.config import settings
from app.core.logging import get_logger

logger = get_logger(__name__)


class PrivacyLevel(Enum):
    """Privacy protection levels"""
    LOW = "low"          # ε=10.0, δ=1e-3
    MEDIUM = "medium"    # ε=1.0, δ=1e-5  
    HIGH = "high"        # ε=0.1, δ=1e-6
    MAXIMUM = "maximum"  # ε=0.01, δ=1e-8


@dataclass
class PrivacyBudget:
    """Privacy budget management"""
    epsilon: float
    delta: float
    spent_epsilon: float = 0.0
    spent_delta: float = 0.0
    operations: List[Dict[str, Any]] = None
    
    def __post_init__(self):
        if self.operations is None:
            self.operations = []
    
    def can_spend(self, epsilon: float, delta: float) -> bool:
        """Check if budget allows spending"""
        return (
            self.spent_epsilon + epsilon <= self.epsilon and
            self.spent_delta + delta <= self.delta
        )
    
    def spend(self, epsilon: float, delta: float, operation: str) -> bool:
        """Spend privacy budget"""
        if not self.can_spend(epsilon, delta):
            return False
        
        self.spent_epsilon += epsilon
        self.spent_delta += delta
        self.operations.append({
            "operation": operation,
            "epsilon": epsilon,
            "delta": delta,
            "timestamp": datetime.now().isoformat()
        })
        return True
    
    def remaining_budget(self) -> Tuple[float, float]:
        """Get remaining privacy budget"""
        return (
            self.epsilon - self.spent_epsilon,
            self.delta - self.spent_delta
        )


class PrivacyEngine:
    """Advanced privacy protection engine"""
    
    def __init__(self):
        self.privacy_levels = {
            PrivacyLevel.LOW: PrivacyBudget(10.0, 1e-3),
            PrivacyLevel.MEDIUM: PrivacyBudget(1.0, 1e-5),
            PrivacyLevel.HIGH: PrivacyBudget(0.1, 1e-6),
            PrivacyLevel.MAXIMUM: PrivacyBudget(0.01, 1e-8)
        }
        self.accountant = BudgetAccountant() if DIFFPRIVLIB_AVAILABLE else None
        self.opendp_available = OPENDP_AVAILABLE
        self.diffprivlib_available = DIFFPRIVLIB_AVAILABLE
        
    async def apply_differential_privacy(
        self,
        data: pd.DataFrame,
        privacy_level: str,
        schema_analysis: Dict[str, Any],
        custom_epsilon: Optional[float] = None,
        custom_delta: Optional[float] = None
    ) -> pd.DataFrame:
        """Apply differential privacy to synthetic data"""
        
        level = PrivacyLevel(privacy_level)
        budget = self.privacy_levels[level]
        
        if custom_epsilon is not None:
            budget.epsilon = custom_epsilon
        if custom_delta is not None:
            budget.delta = custom_delta
        
        logger.info(
            "Applying differential privacy",
            privacy_level=privacy_level,
            epsilon=budget.epsilon,
            delta=budget.delta,
            rows=len(data),
            columns=len(data.columns)
        )
        
        protected_data = data.copy()
        
        # Apply column-specific privacy protection
        for column in data.columns:
            column_info = self._get_column_info(column, schema_analysis)
            
            if column_info.get("privacy_sensitive", False):
                protected_data[column] = await self._protect_sensitive_column(
                    data[column], column_info, budget
                )
            elif column_info.get("data_type") == "numerical":
                protected_data[column] = await self._protect_numerical_column(
                    data[column], column_info, budget
                )
            elif column_info.get("data_type") == "categorical":
                protected_data[column] = await self._protect_categorical_column(
                    data[column], column_info, budget
                )
        
        # Add global noise if budget allows
        remaining_eps, remaining_delta = budget.remaining_budget()
        if remaining_eps > 0.01 and remaining_delta > 1e-8:
            protected_data = await self._add_global_noise(
                protected_data, remaining_eps * 0.1, remaining_delta * 0.1
            )
        
        logger.info(
            "Differential privacy applied successfully",
            budget_used_epsilon=budget.spent_epsilon,
            budget_used_delta=budget.spent_delta,
            remaining_epsilon=remaining_eps,
            remaining_delta=remaining_delta
        )
        
        return protected_data
    
    async def _protect_sensitive_column(
        self,
        column_data: pd.Series,
        column_info: Dict[str, Any],
        budget: PrivacyBudget
    ) -> pd.Series:
        """Protect sensitive columns with strong privacy guarantees"""
        
        privacy_category = column_info.get("privacy_category", "unknown")
        
        if privacy_category == "PII":
            # Apply strong noise to PII data
            return await self._apply_laplace_noise(
                column_data, epsilon=0.1, sensitivity=1.0, budget=budget
            )
        elif privacy_category == "financial":
            # Financial data needs careful protection
            return await self._apply_gaussian_noise(
                column_data, epsilon=0.2, delta=1e-6, sensitivity=1000.0, budget=budget
            )
        elif privacy_category == "health":
            # Health data requires maximum protection
            return await self._apply_exponential_mechanism(
                column_data, epsilon=0.05, sensitivity=1.0, budget=budget
            )
        else:
            # General sensitive data
            return await self._apply_laplace_noise(
                column_data, epsilon=0.2, sensitivity=1.0, budget=budget
            )
    
    async def _protect_numerical_column(
        self,
        column_data: pd.Series,
        column_info: Dict[str, Any],
        budget: PrivacyBudget
    ) -> pd.Series:
        """Protect numerical columns with appropriate noise"""
        
        # Determine sensitivity based on column range
        data_range = column_data.max() - column_data.min()
        sensitivity = min(data_range * 0.01, 10.0)  # Cap sensitivity
        
        # Use Gaussian mechanism for numerical data
        epsilon = min(0.5, budget.epsilon * 0.1)
        delta = min(1e-6, budget.delta * 0.1)
        
        return await self._apply_gaussian_noise(
            column_data, epsilon, delta, sensitivity, budget
        )
    
    async def _protect_categorical_column(
        self,
        column_data: pd.Series,
        column_info: Dict[str, Any],
        budget: PrivacyBudget
    ) -> pd.Series:
        """Protect categorical columns using randomized response"""
        
        unique_values = column_data.unique()
        
        if len(unique_values) <= 2:
            # Binary categorical - use randomized response
            epsilon = min(1.0, budget.epsilon * 0.2)
            return await self._apply_randomized_response(
                column_data, epsilon, budget
            )
        else:
            # Multi-category - use exponential mechanism
            epsilon = min(0.5, budget.epsilon * 0.15)
            return await self._apply_exponential_mechanism(
                column_data, epsilon, sensitivity=1.0, budget=budget
            )
    
    async def _apply_laplace_noise(
        self,
        data: pd.Series,
        epsilon: float,
        sensitivity: float,
        budget: PrivacyBudget
    ) -> pd.Series:
        """Apply Laplace noise for differential privacy"""
        
        if not budget.spend(epsilon, 0.0, f"laplace_noise_{data.name}"):
            logger.warning("Insufficient privacy budget for Laplace noise")
            return data
        
        # Use OpenDP for Laplace mechanism
        scale = sensitivity / epsilon
        noise = np.random.laplace(0, scale, size=len(data))
        
        # Apply noise while preserving data type constraints
        if data.dtype.kind in ['i', 'f']:  # numeric
            noisy_data = data + noise
            # Clamp to reasonable bounds
            if data.dtype.kind == 'i':
                noisy_data = noisy_data.round().astype(data.dtype)
        else:
            # For non-numeric, apply categorical noise
            noisy_data = data.copy()
        
        return noisy_data
    
    async def _apply_gaussian_noise(
        self,
        data: pd.Series,
        epsilon: float,
        delta: float,
        sensitivity: float,
        budget: PrivacyBudget
    ) -> pd.Series:
        """Apply Gaussian noise for (ε,δ)-differential privacy"""
        
        if not budget.spend(epsilon, delta, f"gaussian_noise_{data.name}"):
            logger.warning("Insufficient privacy budget for Gaussian noise")
            return data
        
        # Calculate noise scale for Gaussian mechanism
        if epsilon == 0:
            return data
        
        c = np.sqrt(2 * np.log(1.25 / delta))
        sigma = c * sensitivity / epsilon
        
        noise = np.random.normal(0, sigma, size=len(data))
        
        if data.dtype.kind in ['i', 'f']:  # numeric
            noisy_data = data + noise
            if data.dtype.kind == 'i':
                noisy_data = noisy_data.round().astype(data.dtype)
        else:
            noisy_data = data.copy()
        
        return noisy_data
    
    async def _apply_randomized_response(
        self,
        data: pd.Series,
        epsilon: float,
        budget: PrivacyBudget
    ) -> pd.Series:
        """Apply randomized response for categorical data"""
        
        if not budget.spend(epsilon, 0.0, f"randomized_response_{data.name}"):
            logger.warning("Insufficient privacy budget for randomized response")
            return data
        
        # Randomized response mechanism
        p = np.exp(epsilon) / (np.exp(epsilon) + 1)
        
        # Create randomized version
        noisy_data = data.copy()
        random_mask = np.random.random(len(data)) > p
        
        if random_mask.any():
            # Flip some values randomly
            unique_vals = data.unique()
            if len(unique_vals) == 2:
                # Binary case
                other_val = unique_vals[1] if data.iloc[0] == unique_vals[0] else unique_vals[0]
                noisy_data.loc[random_mask] = other_val
            else:
                # Multi-category case
                random_choices = np.random.choice(unique_vals, size=random_mask.sum())
                noisy_data.loc[random_mask] = random_choices
        
        return noisy_data
    
    async def _apply_exponential_mechanism(
        self,
        data: pd.Series,
        epsilon: float,
        sensitivity: float,
        budget: PrivacyBudget
    ) -> pd.Series:
        """Apply exponential mechanism for categorical data"""
        
        if not budget.spend(epsilon, 0.0, f"exponential_mechanism_{data.name}"):
            logger.warning("Insufficient privacy budget for exponential mechanism")
            return data
        
        # Exponential mechanism with utility based on frequency
        value_counts = data.value_counts()
        utilities = value_counts.values
        
        # Normalize utilities
        max_utility = utilities.max()
        normalized_utilities = utilities / max_utility
        
        # Exponential weights
        weights = np.exp(epsilon * normalized_utilities / (2 * sensitivity))
        probabilities = weights / weights.sum()
        
        # Sample new values based on exponential mechanism
        noisy_data = data.copy()
        unique_vals = value_counts.index.values
        
        for i in range(len(data)):
            if np.random.random() < 0.1:  # 10% chance to change
                new_val = np.random.choice(unique_vals, p=probabilities)
                noisy_data.iloc[i] = new_val
        
        return noisy_data
    
    async def _add_global_noise(
        self,
        data: pd.DataFrame,
        epsilon: float,
        delta: float
    ) -> pd.DataFrame:
        """Add global noise to protect against correlation attacks"""
        
        # Add small amount of correlated noise across columns
        correlation_noise = np.random.multivariate_normal(
            mean=np.zeros(len(data.columns)),
            cov=np.eye(len(data.columns)) * 0.01,
            size=len(data)
        )
        
        noisy_data = data.copy()
        numeric_columns = data.select_dtypes(include=[np.number]).columns
        
        for i, col in enumerate(numeric_columns):
            if i < correlation_noise.shape[1]:
                noisy_data[col] = data[col] + correlation_noise[:, i]
        
        return noisy_data
    
    def _get_column_info(
        self, 
        column_name: str, 
        schema_analysis: Dict[str, Any]
    ) -> Dict[str, Any]:
        """Get column information from schema analysis"""
        
        # Extract column info from schema analysis
        columns_info = schema_analysis.get("columns", [])
        
        for col_info in columns_info:
            if col_info.get("name") == column_name:
                return col_info
        
        # Default column info if not found
        return {
            "name": column_name,
            "data_type": "unknown",
            "privacy_sensitive": False,
            "privacy_category": "general"
        }
    
    async def generate_privacy_report(
        self,
        original_data: pd.DataFrame,
        protected_data: pd.DataFrame,
        privacy_level: str,
        budget_used: PrivacyBudget
    ) -> Dict[str, Any]:
        """Generate comprehensive privacy protection report"""
        
        # Calculate privacy metrics
        privacy_metrics = {
            "privacy_level": privacy_level,
            "epsilon_used": budget_used.spent_epsilon,
            "delta_used": budget_used.spent_delta,
            "operations_count": len(budget_used.operations),
            "data_utility": await self._calculate_data_utility(original_data, protected_data),
            "privacy_risk": await self._assess_privacy_risk(protected_data, budget_used),
            "compliance_status": await self._check_compliance(budget_used),
            "recommendations": await self._generate_recommendations(budget_used)
        }
        
        return privacy_metrics
    
    async def _calculate_data_utility(
        self,
        original: pd.DataFrame,
        protected: pd.DataFrame
    ) -> float:
        """Calculate utility preservation after privacy protection"""
        
        try:
            # Calculate various utility metrics
            numeric_cols = original.select_dtypes(include=[np.number]).columns
            
            if len(numeric_cols) == 0:
                return 1.0  # No numeric columns to compare
            
            # Mean squared error for numeric columns
            mse_scores = []
            for col in numeric_cols:
                if col in protected.columns:
                    mse = np.mean((original[col] - protected[col]) ** 2)
                    original_var = np.var(original[col])
                    relative_mse = mse / original_var if original_var > 0 else 0
                    mse_scores.append(1 - min(relative_mse, 1.0))
            
            return np.mean(mse_scores) if mse_scores else 1.0
            
        except Exception as e:
            logger.warning("Utility calculation failed", error=str(e))
            return 0.8  # Default utility score
    
    async def _assess_privacy_risk(
        self,
        protected_data: pd.DataFrame,
        budget: PrivacyBudget
    ) -> str:
        """Assess privacy risk level"""
        
        total_epsilon = budget.spent_epsilon
        
        if total_epsilon <= 0.1:
            return "very_low"
        elif total_epsilon <= 1.0:
            return "low"
        elif total_epsilon <= 5.0:
            return "medium"
        elif total_epsilon <= 10.0:
            return "high"
        else:
            return "very_high"
    
    async def _check_compliance(self, budget: PrivacyBudget) -> Dict[str, bool]:
        """Check compliance with privacy regulations"""
        
        return {
            "gdpr_compliant": budget.spent_epsilon <= 1.0,
            "ccpa_compliant": budget.spent_epsilon <= 5.0,
            "hipaa_ready": budget.spent_epsilon <= 0.1,
            "ferpa_compliant": budget.spent_epsilon <= 1.0
        }
    
    async def _generate_recommendations(
        self, 
        budget: PrivacyBudget
    ) -> List[str]:
        """Generate privacy recommendations"""
        
        recommendations = []
        
        remaining_eps, remaining_delta = budget.remaining_budget()
        
        if remaining_eps < 0.1:
            recommendations.append(
                "Privacy budget nearly exhausted. Consider using higher epsilon for future operations."
            )
        
        if budget.spent_epsilon > 5.0:
            recommendations.append(
                "High epsilon usage detected. Consider stronger privacy protection."
            )
        
        if len(budget.operations) > 10:
            recommendations.append(
                "Many privacy operations performed. Consider consolidating operations."
            )
        
        return recommendations 

File: backend-go/internal/privacy/privacy_engine.go
package privacy

import (
	"crypto/rand"
	"fmt"
	"math"
	"math/big"
	"time"
)

// PrivacyLevel represents different privacy protection levels
type PrivacyLevel string

const (
	PrivacyLevelLow     PrivacyLevel = "low"
	PrivacyLevelMedium  PrivacyLevel = "medium"
	PrivacyLevelHigh    PrivacyLevel = "high"
	PrivacyLevelMaximum PrivacyLevel = "maximum"
)

// PrivacyBudget manages privacy budget for differential privacy
type PrivacyBudget struct {
	Epsilon      float64     `json:"epsilon"`
	Delta        float64     `json:"delta"`
	SpentEpsilon float64     `json:"spent_epsilon"`
	SpentDelta   float64     `json:"spent_delta"`
	Operations   []Operation `json:"operations"`
}

// Operation represents a privacy operation
type Operation struct {
	Operation string    `json:"operation"`
	Epsilon   float64   `json:"epsilon"`
	Delta     float64   `json:"delta"`
	Timestamp time.Time `json:"timestamp"`
}

// PrivacyEngine handles differential privacy and data protection
type PrivacyEngine struct {
	privacyLevels map[PrivacyLevel]*PrivacyBudget
}

// NewPrivacyEngine creates a new privacy engine
func NewPrivacyEngine() *PrivacyEngine {
	return &PrivacyEngine{
		privacyLevels: map[PrivacyLevel]*PrivacyBudget{
			PrivacyLevelLow: {
				Epsilon: 10.0,
				Delta:   1e-3,
			},
			PrivacyLevelMedium: {
				Epsilon: 1.0,
				Delta:   1e-5,
			},
			PrivacyLevelHigh: {
				Epsilon: 0.1,
				Delta:   1e-6,
			},
			PrivacyLevelMaximum: {
				Epsilon: 0.01,
				Delta:   1e-8,
			},
		},
	}
}

// ApplyDifferentialPrivacy applies differential privacy to data
func (p *PrivacyEngine) ApplyDifferentialPrivacy(data []map[string]interface{}, privacyLevel PrivacyLevel, schema map[string]interface{}) ([]map[string]interface{}, error) {
	budget := p.privacyLevels[privacyLevel]
	if budget == nil {
		return nil, fmt.Errorf("invalid privacy level: %s", privacyLevel)
	}

	protectedData := make([]map[string]interface{}, len(data))

	for i, row := range data {
		protectedRow := make(map[string]interface{})

		for key, value := range row {
			columnInfo := p.getColumnInfo(key, schema)

			if columnInfo.PrivacySensitive {
				protectedValue, err := p.protectSensitiveColumn(value, columnInfo, budget)
				if err != nil {
					return nil, fmt.Errorf("failed to protect sensitive column %s: %w", key, err)
				}
				protectedRow[key] = protectedValue
			} else if columnInfo.DataType == "numerical" {
				protectedValue, err := p.protectNumericalColumn(value, columnInfo, budget)
				if err != nil {
					return nil, fmt.Errorf("failed to protect numerical column %s: %w", key, err)
				}
				protectedRow[key] = protectedValue
			} else if columnInfo.DataType == "categorical" {
				protectedValue, err := p.protectCategoricalColumn(value, columnInfo, budget)
				if err != nil {
					return nil, fmt.Errorf("failed to protect categorical column %s: %w", key, err)
				}
				protectedRow[key] = protectedValue
			} else {
				protectedRow[key] = value
			}
		}

		protectedData[i] = protectedRow
	}

	return protectedData, nil
}

// ColumnInfo represents information about a data column
type ColumnInfo struct {
	Name             string        `json:"name"`
	DataType         string        `json:"data_type"`
	PrivacySensitive bool          `json:"privacy_sensitive"`
	PrivacyCategory  string        `json:"privacy_category"`
	MinValue         float64       `json:"min_value"`
	MaxValue         float64       `json:"max_value"`
	UniqueValues     []interface{} `json:"unique_values"`
}

// getColumnInfo extracts column information from schema
func (p *PrivacyEngine) getColumnInfo(columnName string, schema map[string]interface{}) ColumnInfo {
	columns, ok := schema["columns"].([]interface{})
	if !ok {
		return ColumnInfo{
			Name:             columnName,
			DataType:         "unknown",
			PrivacySensitive: false,
			PrivacyCategory:  "general",
		}
	}

	for _, col := range columns {
		colMap, ok := col.(map[string]interface{})
		if !ok {
			continue
		}

		if colMap["name"] == columnName {
			return ColumnInfo{
				Name:             getString(colMap, "name", columnName),
				DataType:         getString(colMap, "data_type", "unknown"),
				PrivacySensitive: getBool(colMap, "privacy_sensitive", false),
				PrivacyCategory:  getString(colMap, "privacy_category", "general"),
				MinValue:         getFloat64(colMap, "min_value", 0),
				MaxValue:         getFloat64(colMap, "max_value", 100),
				UniqueValues:     getSlice(colMap, "unique_values", []interface{}{}),
			}
		}
	}

	return ColumnInfo{
		Name:             columnName,
		DataType:         "unknown",
		PrivacySensitive: false,
		PrivacyCategory:  "general",
	}
}

// protectSensitiveColumn applies strong privacy protection to sensitive columns
func (p *PrivacyEngine) protectSensitiveColumn(value interface{}, columnInfo ColumnInfo, budget *PrivacyBudget) (interface{}, error) {
	privacyCategory := columnInfo.PrivacyCategory

	switch privacyCategory {
	case "PII":
		return p.applyLaplaceNoise(value, 0.1, 1.0, budget)
	case "financial":
		return p.applyGaussianNoise(value, 0.2, 1e-6, 1000.0, budget)
	case "health":
		return p.applyExponentialMechanism(value, 0.05, 1.0, budget)
	default:
		return p.applyLaplaceNoise(value, 0.2, 1.0, budget)
	}
}

// protectNumericalColumn applies appropriate noise to numerical columns
func (p *PrivacyEngine) protectNumericalColumn(value interface{}, columnInfo ColumnInfo, budget *PrivacyBudget) (interface{}, error) {
	// Determine sensitivity based on column range
	dataRange := columnInfo.MaxValue - columnInfo.MinValue
	sensitivity := math.Min(dataRange*0.01, 10.0) // Cap sensitivity

	// Use Gaussian mechanism for numerical data
	epsilon := math.Min(0.5, budget.Epsilon*0.1)
	delta := math.Min(1e-6, budget.Delta*0.1)

	return p.applyGaussianNoise(value, epsilon, delta, sensitivity, budget)
}

// protectCategoricalColumn applies randomized response to categorical columns
func (p *PrivacyEngine) protectCategoricalColumn(value interface{}, columnInfo ColumnInfo, budget *PrivacyBudget) (interface{}, error) {
	uniqueValues := columnInfo.UniqueValues

	if len(uniqueValues) <= 2 {
		// Binary categorical - use randomized response
		epsilon := math.Min(1.0, budget.Epsilon*0.2)
		return p.applyRandomizedResponse(value, epsilon, uniqueValues, budget)
	} else {
		// Multi-category - use exponential mechanism
		epsilon := math.Min(0.5, budget.Epsilon*0.15)
		return p.applyExponentialMechanism(value, epsilon, 1.0, budget)
	}
}

// applyLaplaceNoise applies Laplace noise for differential privacy
func (p *PrivacyEngine) applyLaplaceNoise(value interface{}, epsilon, sensitivity float64, budget *PrivacyBudget) (interface{}, error) {
	if !budget.canSpend(epsilon, 0.0) {
		return value, fmt.Errorf("insufficient privacy budget for Laplace noise")
	}

	budget.spend(epsilon, 0.0, fmt.Sprintf("laplace_noise_%s", "column"))

	// Generate Laplace noise
	scale := sensitivity / epsilon
	noise := p.generateLaplaceNoise(scale)

	// Apply noise based on value type
	switch v := value.(type) {
	case float64:
		return v + noise, nil
	case int:
		return int(float64(v) + noise), nil
	case int64:
		return int64(float64(v) + noise), nil
	default:
		return value, nil
	}
}

// applyGaussianNoise applies Gaussian noise for (ε,δ)-differential privacy
func (p *PrivacyEngine) applyGaussianNoise(value interface{}, epsilon, delta, sensitivity float64, budget *PrivacyBudget) (interface{}, error) {
	if !budget.canSpend(epsilon, delta) {
		return value, fmt.Errorf("insufficient privacy budget for Gaussian noise")
	}

	budget.spend(epsilon, delta, fmt.Sprintf("gaussian_noise_%s", "column"))

	// Calculate noise scale for Gaussian mechanism
	if epsilon == 0 {
		return value, nil
	}

	c := math.Sqrt(2 * math.Log(1.25/delta))
	sigma := c * sensitivity / epsilon

	noise := p.generateGaussianNoise(sigma)

	// Apply noise based on value type
	switch v := value.(type) {
	case float64:
		return v + noise, nil
	case int:
		return int(float64(v) + noise), nil
	case int64:
		return int64(float64(v) + noise), nil
	default:
		return value, nil
	}
}

// applyRandomizedResponse applies randomized response for categorical data
func (p *PrivacyEngine) applyRandomizedResponse(value interface{}, epsilon float64, uniqueValues []interface{}, budget *PrivacyBudget) (interface{}, error) {
	if !budget.canSpend(epsilon, 0.0) {
		return value, fmt.Errorf("insufficient privacy budget for randomized response")
	}

	budget.spend(epsilon, 0.0, fmt.Sprintf("randomized_response_%s", "column"))

	// Randomized response mechanism
	p := math.Exp(epsilon) / (math.Exp(epsilon) + 1)

	// Random decision
	random, err := p.generateRandomFloat()
	if err != nil {
		return value, err
	}

	if random > p {
		// Flip value
		if len(uniqueValues) == 2 {
			// Binary case
			if value == uniqueValues[0] {
				return uniqueValues[1], nil
			} else {
				return uniqueValues[0], nil
			}
		} else {
			// Multi-category case
			randomIndex, err := p.generateRandomInt(len(uniqueValues))
			if err != nil {
				return value, err
			}
			return uniqueValues[randomIndex], nil
		}
	}

	return value, nil
}

// applyExponentialMechanism applies exponential mechanism for categorical data
func (p *PrivacyEngine) applyExponentialMechanism(value interface{}, epsilon, sensitivity float64, budget *PrivacyBudget) (interface{}, error) {
	if !budget.canSpend(epsilon, 0.0) {
		return value, fmt.Errorf("insufficient privacy budget for exponential mechanism")
	}

	budget.spend(epsilon, 0.0, fmt.Sprintf("exponential_mechanism_%s", "column"))

	// For simplicity, return original value with small probability of change
	random, err := p.generateRandomFloat()
	if err != nil {
		return value, err
	}

	if random < 0.1 { // 10% chance to change
		// This is a simplified implementation
		// In practice, you'd implement the full exponential mechanism
		return value, nil
	}

	return value, nil
}

// GeneratePrivacyReport generates a comprehensive privacy protection report
func (p *PrivacyEngine) GeneratePrivacyReport(originalData, protectedData []map[string]interface{}, privacyLevel PrivacyLevel, budget *PrivacyBudget) map[string]interface{} {
	// Calculate privacy metrics
	privacyMetrics := map[string]interface{}{
		"privacy_level":     privacyLevel,
		"epsilon_used":      budget.SpentEpsilon,
		"delta_used":        budget.SpentDelta,
		"operations_count":  len(budget.Operations),
		"data_utility":      p.calculateDataUtility(originalData, protectedData),
		"privacy_risk":      p.assessPrivacyRisk(budget),
		"compliance_status": p.checkCompliance(budget),
		"recommendations":   p.generateRecommendations(budget),
	}

	return privacyMetrics
}

// calculateDataUtility calculates utility preservation after privacy protection
func (p *PrivacyEngine) calculateDataUtility(original, protected []map[string]interface{}) float64 {
	if len(original) != len(protected) {
		return 0.0
	}

	// Simplified utility calculation
	// In practice, you'd implement more sophisticated metrics
	return 0.85 // Default utility score
}

// assessPrivacyRisk assesses privacy risk level
func (p *PrivacyEngine) assessPrivacyRisk(budget *PrivacyBudget) string {
	totalEpsilon := budget.SpentEpsilon

	if totalEpsilon <= 0.1 {
		return "very_low"
	} else if totalEpsilon <= 1.0 {
		return "low"
	} else if totalEpsilon <= 5.0 {
		return "medium"
	} else if totalEpsilon <= 10.0 {
		return "high"
	} else {
		return "very_high"
	}
}

// checkCompliance checks compliance with privacy regulations
func (p *PrivacyEngine) checkCompliance(budget *PrivacyBudget) map[string]bool {
	return map[string]bool{
		"gdpr_compliant":  budget.SpentEpsilon <= 1.0,
		"ccpa_compliant":  budget.SpentEpsilon <= 5.0,
		"hipaa_ready":     budget.SpentEpsilon <= 0.1,
		"ferpa_compliant": budget.SpentEpsilon <= 1.0,
	}
}

// generateRecommendations generates privacy recommendations
func (p *PrivacyEngine) generateRecommendations(budget *PrivacyBudget) []string {
	var recommendations []string

	remainingEps := budget.Epsilon - budget.SpentEpsilon
	remainingDelta := budget.Delta - budget.SpentDelta

	if remainingEps < 0.1 {
		recommendations = append(recommendations, "Privacy budget nearly exhausted. Consider using higher epsilon for future operations.")
	}

	if budget.SpentEpsilon > 5.0 {
		recommendations = append(recommendations, "High epsilon usage detected. Consider stronger privacy protection.")
	}

	if len(budget.Operations) > 10 {
		recommendations = append(recommendations, "Many privacy operations performed. Consider consolidating operations.")
	}

	return recommendations
}

// Helper methods for budget management
func (b *PrivacyBudget) canSpend(epsilon, delta float64) bool {
	return b.SpentEpsilon+epsilon <= b.Epsilon && b.SpentDelta+delta <= b.Delta
}

func (b *PrivacyBudget) spend(epsilon, delta float64, operation string) bool {
	if !b.canSpend(epsilon, delta) {
		return false
	}

	b.SpentEpsilon += epsilon
	b.SpentDelta += delta
	b.Operations = append(b.Operations, Operation{
		Operation: operation,
		Epsilon:   epsilon,
		Delta:     delta,
		Timestamp: time.Now(),
	})

	return true
}

// Helper methods for random number generation
func (p *PrivacyEngine) generateLaplaceNoise(scale float64) float64 {
	// Generate uniform random number in [0, 1)
	u, _ := p.generateRandomFloat()

	// Transform to Laplace distribution
	if u < 0.5 {
		return scale * math.Log(2*u)
	} else {
		return -scale * math.Log(2*(1-u))
	}
}

func (p *PrivacyEngine) generateGaussianNoise(sigma float64) float64 {
	// Box-Muller transform for Gaussian distribution
	u1, _ := p.generateRandomFloat()
	u2, _ := p.generateRandomFloat()

	z0 := math.Sqrt(-2*math.Log(u1)) * math.Cos(2*math.Pi*u2)
	return sigma * z0
}

func (p *PrivacyEngine) generateRandomFloat() (float64, error) {
	n, err := rand.Int(rand.Reader, big.NewInt(1000000))
	if err != nil {
		return 0, err
	}
	return float64(n.Int64()) / 1000000.0, nil
}

func (p *PrivacyEngine) generateRandomInt(max int) (int, error) {
	n, err := rand.Int(rand.Reader, big.NewInt(int64(max)))
	if err != nil {
		return 0, err
	}
	return int(n.Int64()), nil
}

// Helper functions for type conversion
func getString(m map[string]interface{}, key, defaultValue string) string {
	if v, ok := m[key].(string); ok {
		return v
	}
	return defaultValue
}

func getBool(m map[string]interface{}, key string, defaultValue bool) bool {
	if v, ok := m[key].(bool); ok {
		return v
	}
	return defaultValue
}

func getFloat64(m map[string]interface{}, key string, defaultValue float64) float64 {
	if v, ok := m[key].(float64); ok {
		return v
	}
	return defaultValue
}

func getSlice(m map[string]interface{}, key string, defaultValue []interface{}) []interface{} {
	if v, ok := m[key].([]interface{}); ok {
		return v
	}
	return defaultValue
}
